Problem to start minio server in distributed mode after upgrade to 2018-03-16T22:52:12Z

I just upgrade minio to latest version.

```
$ minio version
Version: 2018-03-16T22:52:12Z
Release-Tag: RELEASE.2018-03-16T22-52-12Z
Commit-ID: f23944aed7b7980d4400ac8bc022dabd8ac1c273
```

and after upgrade all 4 server, it's problem to start minio server in distributed mode.

when running command on first server,

```minio server http://node1/mnt/minio/cluster-data http://node2/mnt/minio/cluster-data http://node3/mnt/minio/cluster-data http://node4/mnt/minio/cluster-data```

server start and `"Waiting for a minimum of 2 disks to come online"`

when I run same command on other servers, I see error
```
Trace: 1: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/server-main.go:247:cmd.serverMain()
       2: /q/.q/sources/gopath/src/github.com/minio/minio/vendor/github.com/minio/cli/app.go:499:cli.HandleAction()
       3: /q/.q/sources/gopath/src/github.com/minio/minio/vendor/github.com/minio/cli/command.go:214:cli.Command.Run()
       4: /q/.q/sources/gopath/src/github.com/minio/minio/vendor/github.com/minio/cli/app.go:260:cli.(*App).Run()
       5: /q/.q/sources/gopath/src/github.com/minio/minio/cmd/main.go:155:cmd.Main()
       6: /q/.q/sources/gopath/src/github.com/minio/minio/main.go:71:main.main()
[2018-03-18T13:50:14.132962493Z] [ERROR] Initializing object layer failed (http://node3:9000/mnt/minio/cluster-data format error: Disk ID  not found in any disk sets 
[[427bef43-e2bf-4e79-a7bd-00b820ba241e c7524dab-2dcd-4747-8dd5-89a593cd3b8d 3bdea66e-5698-477a-bfe2-7d331fff979d faa851b2-d9a1-4cb1-b369-da674846f6b6]])
```

format.json
```json
{"version":"1","format":"xl","xl":{
"version":"3",
"this":"",
"sets":[
["427bef43-e2bf-4e79-a7bd-00b820ba241e",
"c7524dab-2dcd-4747-8dd5-89a593cd3b8d",
"3bdea66e-5698-477a-bfe2-7d331fff979d",
"faa851b2-d9a1-4cb1-b369-da674846f6b6"]],
"distributionAlgo":"CRCMOD"}}
```

## Possible Solution
Thank for help @y4m4 on slack
Manual setup UUID per server in format.json and minio server start working again.

server1:
```json
{"version":"1","format":"xl","xl":{
"version":"3","this":"427bef43-e2bf-4e79-a7bd-00b820ba241e",
"sets":[["427bef43-e2bf-4e79-a7bd-00b820ba241e","c7524dab-2dcd-4747-8dd5-89a593cd3b8d","3bdea66e-5698-477a-bfe2-7d331fff979d","faa851b2-d9a1-4cb1-b369-da674846f6b6"]],"distributionAlgo":"CRCMOD"}}
```

server2:
```json
{"version":"1","format":"xl","xl":{
"version":"3","this":"c7524dab-2dcd-4747-8dd5-89a593cd3b8d",
"sets":[["427bef43-e2bf-4e79-a7bd-00b820ba241e","c7524dab-2dcd-4747-8dd5-89a593cd3b8d","3bdea66e-5698-477a-bfe2-7d331fff979d","faa851b2-d9a1-4cb1-b369-da674846f6b6"]],"distributionAlgo":"CRCMOD"}}
```
server3:
```json
{"version":"1","format":"xl","xl":{
"version":"3","this":"3bdea66e-5698-477a-bfe2-7d331fff979d",
"sets":[["427bef43-e2bf-4e79-a7bd-00b820ba241e","c7524dab-2dcd-4747-8dd5-89a593cd3b8d","3bdea66e-5698-477a-bfe2-7d331fff979d","faa851b2-d9a1-4cb1-b369-da674846f6b6"]],"distributionAlgo":"CRCMOD"}}
```
server4:
```json
{"version":"1","format":"xl","xl":{
"version":"3","this":"faa851b2-d9a1-4cb1-b369-da674846f6b6",
"sets":[["427bef43-e2bf-4e79-a7bd-00b820ba241e","c7524dab-2dcd-4747-8dd5-89a593cd3b8d","3bdea66e-5698-477a-bfe2-7d331fff979d","faa851b2-d9a1-4cb1-b369-da674846f6b6"]],"distributionAlgo":"CRCMOD"}}
```

## Environment
Linux  4.13.0-37-generic #42~16.04.1-Ubuntu SMP Wed Mar 7 16:03:28 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
current minio version: 2018-03-16T22:52:12Z
old minio version: 1 month old (I dont know exactly)

--- a/cmd/format-xl.go
+++ b/cmd/format-xl.go
@@ -232,6 +232,7 @@ func formatXLMigrateV1ToV2(export string) error {
 	formatV2.Format = formatBackendXL
 	formatV2.XL.Version = formatXLVersionV2
 	formatV2.XL.DistributionAlgo = formatXLVersionV2DistributionAlgo
+	formatV2.XL.This = formatV1.XL.Disk
 	formatV2.XL.Sets = make([][]string, 1)
 	formatV2.XL.Sets[0] = make([]string, len(formatV1.XL.JBOD))
 	copy(formatV2.XL.Sets[0], formatV1.XL.JBOD)
@@ -600,6 +601,55 @@ func initStorageDisks(endpoints EndpointList) ([]StorageAPI, error) {
 		storageDisks[index] = storage
 	}
 	return storageDisks, nil
+}
+
+// formatXLV3ThisEmpty - find out if '.This' field is empty
+// in any of the input `formats`, if yes return true.
+func formatXLV3ThisEmpty(formats []*formatXLV3) bool {
+	for _, format := range formats {
+		if format == nil {
+			continue
+		}
+		// NOTE: This code is specifically needed when migrating version
+		// V1 to V2 to V3, in a scenario such as this we only need to handle
+		// single sets since we never used to support multiple sets in releases
+		// with V1 format version.
+		if len(format.XL.Sets) > 1 {
+			continue
+		}
+		if format.XL.This == "" {
+			return true
+		}
+	}
+	return false
+}
+
+// fixFormatXLV3 - fix format XL configuration on all disks.
+func fixFormatXLV3(endpoints EndpointList, formats []*formatXLV3) error {
+	storageDisks, err := initStorageDisks(endpoints)
+	if err != nil {
+		return err
+	}
+
+	for i, format := range formats {
+		if format == nil || !endpoints[i].IsLocal {
+			continue
+		}
+		// NOTE: This code is specifically needed when migrating version
+		// V1 to V2 to V3, in a scenario such as this we only need to handle
+		// single sets since we never used to support multiple sets in releases
+		// with V1 format version.
+		if len(format.XL.Sets) > 1 {
+			continue
+		}
+		if format.XL.This == "" {
+			formats[i].XL.This = format.XL.Sets[0][i]
+			if err = saveFormatXL(storageDisks[i], formats[i]); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
 }
 
 // initFormatXL - save XL format configuration on all disks.

--- a/cmd/format-xl_test.go
+++ b/cmd/format-xl_test.go
@@ -18,10 +18,12 @@ package cmd
 
 import (
 	"encoding/json"
-	"fmt"
 	"io/ioutil"
 	"os"
+	"reflect"
 	"testing"
+
+	"github.com/minio/minio/pkg/errors"
 )
 
 // Test get offline/online uuids.
@@ -71,6 +73,73 @@ func TestGetUUIDs(t *testing.T) {
 	}
 	if gotCount != 16 {
 		t.Errorf("Expected offline count '16', got '%d'", gotCount)
+	}
+}
+
+// tests fixFormatXLV3 - fix format.json on all disks.
+func TestFixFormatV3(t *testing.T) {
+	xlDirs, err := getRandomDisks(8)
+	if err != nil {
+		t.Fatal(err)
+	}
+	for _, xlDir := range xlDirs {
+		defer os.RemoveAll(xlDir)
+	}
+	endpoints := mustGetNewEndpointList(xlDirs...)
+
+	format := newFormatXLV3(1, 8)
+	formats := make([]*formatXLV3, 8)
+
+	for j := 0; j < 8; j++ {
+		newFormat := *format
+		newFormat.XL.This = format.XL.Sets[0][j]
+		formats[j] = &newFormat
+	}
+
+	if err = initFormatXLMetaVolume(endpoints, formats); err != nil {
+		t.Fatal(err)
+	}
+
+	formats[1] = nil
+	expThis := formats[2].XL.This
+	formats[2].XL.This = ""
+	if err := fixFormatXLV3(endpoints, formats); err != nil {
+		t.Fatal(err)
+	}
+	newFormats, errs := loadFormatXLAll(endpoints)
+	for _, err := range errs {
+		if err != nil && errors.Cause(err) != errUnformattedDisk {
+			t.Fatal(err)
+		}
+	}
+	gotThis := newFormats[2].XL.This
+	if expThis != gotThis {
+		t.Fatalf("expected uuid %s, got %s", expThis, gotThis)
+	}
+}
+
+// tests formatXLV3ThisEmpty conditions.
+func TestFormatXLEmpty(t *testing.T) {
+	format := newFormatXLV3(1, 16)
+	formats := make([]*formatXLV3, 16)
+
+	for j := 0; j < 16; j++ {
+		newFormat := *format
+		newFormat.XL.This = format.XL.Sets[0][j]
+		formats[j] = &newFormat
+	}
+
+	// empty format to indicate disk not found, but this
+	// empty should return false.
+	formats[0] = nil
+
+	if ok := formatXLV3ThisEmpty(formats); ok {
+		t.Fatalf("expected value false, got %t", ok)
+	}
+
+	formats[2].XL.This = ""
+	if ok := formatXLV3ThisEmpty(formats); !ok {
+		t.Fatalf("expected value true, got %t", ok)
 	}
 }
 
@@ -201,7 +270,25 @@ func TestFormatXLMigrate(t *testing.T) {
 		t.Fatal(err)
 	}
 	if migratedVersion != formatXLVersionV3 {
-		t.Fatal(fmt.Sprintf("expected version: %s, got: %s", formatXLVersionV3, migratedVersion))
+		t.Fatalf("expected version: %s, got: %s", formatXLVersionV3, migratedVersion)
+	}
+
+	b, err = ioutil.ReadFile(pathJoin(rootPath, minioMetaBucket, formatConfigFile))
+	if err != nil {
+		t.Fatal(err)
+	}
+	formatV3 := &formatXLV3{}
+	if err = json.Unmarshal(b, formatV3); err != nil {
+		t.Fatal(err)
+	}
+	if formatV3.XL.This != m.XL.Disk {
+		t.Fatalf("expected disk uuid: %s, got: %s", m.XL.Disk, formatV3.XL.This)
+	}
+	if len(formatV3.XL.Sets) != 1 {
+		t.Fatalf("expected single set after migrating from v1 to v3, but found %d", len(formatV3.XL.Sets))
+	}
+	if !reflect.DeepEqual(formatV3.XL.Sets[0], m.XL.JBOD) {
+		t.Fatalf("expected disk uuid: %v, got: %v", m.XL.JBOD, formatV3.XL.Sets[0])
 	}
 
 	m = &formatXLV1{}

--- a/cmd/prepare-storage.go
+++ b/cmd/prepare-storage.go
@@ -147,6 +147,20 @@ func waitForFormatXL(firstDisk bool, endpoints EndpointList, setCount, disksPerS
 				return initFormatXL(endpoints, setCount, disksPerSet)
 			}
 
+			// Following function is added to fix a regressions which was introduced
+			// in release RELEASE.2018-03-16T22-52-12Z after migrating v1 to v2 to v3.
+			// This migration failed to capture '.This' field properly which indicates
+			// the disk UUID association. Below function is called to handle and fix
+			// this regression, for more info refer https://github.com/minio/minio/issues/5667
+			if err = fixFormatXLV3(endpoints, formatConfigs); err != nil {
+				return nil, err
+			}
+
+			// If any of the .This field is still empty we wait them to be fixed.
+			if formatXLV3ThisEmpty(formatConfigs) {
+				continue
+			}
+
 			format, err = getFormatXLInQuorum(formatConfigs)
 			if err == nil {
 				for i := range formatConfigs {

